include/cuda/std/detail/libcxx/include/atomic

```diff
@@ -1243,269 +1243,282 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
         __cxx_atomic_try_wait_slow(__a, __val, __order);
 }
 
-// general atomic<T>/atomic_ref<T>
-template <class _Tp, int _Sco = 0, bool = (is_integral<_Tp>::value || is_floating_point<_Tp>::value) && !is_same<_Tp, bool>::value>
-struct __atomic_base {
-    mutable __cxx_atomic_impl<_Tp, _Sco> __a_;
+template <class _Tp, typename _Storage>
+struct __atomic_base_storage {
+    mutable _Storage __a_;
 
-    __atomic_base() = default;
-    __atomic_base(const __atomic_base&) = delete;
-    __atomic_base(__atomic_base&&) = delete;
+    __atomic_base_storage() = default;
+    __atomic_base_storage(const __atomic_base_storage&) = default;
```

Comment:
I am having some troubles correctly parsing the difference between the various classes. AFAICT the only difference is whether the special member functions are deleted / defaulted.

Given that the implementation of the classes is considerable, would it make sense to just derive from a single base class that does this for us, like optional does